xla/service/gpu/cudnn_fused_mha_rewriter.cc

```diff
@@ -1530,40 +1739,52 @@ StatusOr<bool> CudnnFusedMHARewriter::Run(
     }
     for (HloInstruction* instr : comp->MakeInstructionPostOrder()) {
       bool v_transposed = false;
+      bool changed = false;
       MatchFwdResult matched_result =
           MatchFwdMHAPatternsForCanonicalization(instr);
       if (!matched_result.has_match) {
         continue;
       }
+      // flash attention TODO: dont fuse fwd if bwd is not supported
       // We check the validity of bmms here before canonicalization so we don't
       // modify the graph if mha fusion is not possible
+      // Relax 512 constraint if it is flash attention
       TF_ASSIGN_OR_RETURN(
           bool is_mha_module_supported,
           IsMHABlockSupported(
               matched_result.matched_bmm_1, matched_result.matched_bmm_2,
               matched_result.need_canonicalization, matched_result.is_training,
+              matched_result.is_causal_mask, matched_result.is_flash_attention,
               matched_result.matched_custom_call_name, debug_options));
+
       if (!is_mha_module_supported) continue;
+      // We make sure no attention block is matched and replaced twice here
+      if (matched_bmm1.find(matched_result.matched_bmm_1) != matched_bmm1.end())
+        continue;
+      matched_bmm1.insert(matched_result.matched_bmm_1);
```

Comment:
This can be simplified to:

```
if (!matched_bmm1.insert(matched_result.matched_bmm_1).second) {
        continue;
}
```

Note that our style guide also requires using {} around if blocks if the block is not in the same line as the if. So alternatively you can put the continue also in the same line (if it still fits in 80 characters)